crates/bevy_sprite/src/render/mod.rs

```diff
@@ -875,3 +891,96 @@ impl<P: PhaseItem> RenderCommand<P> for DrawSpriteBatch {
+            if quad_ratio > texture_ratio {
+                let scale = Vec2::new(1., -texture_ratio / quad_ratio);
+                let offset = (1.0 - scale) * 0.5;
+                *uv_offset_scale = Vec4::new(offset.x, scale.y.abs(), scale.x, scale.y);
+            } else {
+                let scale = Vec2::new(quad_ratio / texture_ratio, -1.);
+                let offset = (1.0 - scale) * 0.5;
+                *uv_offset_scale = Vec4::new(0.0, offset.y, scale.x, scale.y);
+            }
+        }
+        ScalingMode::FillEnd => {
+            if quad_ratio > texture_ratio {
+                let scale = Vec2::new(1., -texture_ratio / quad_ratio);
+                let offset = (1.0 - scale) * 0.5;
+                *uv_offset_scale = Vec4::new(offset.x, 1.0, scale.x, scale.y);
+            } else {
+                let scale = Vec2::new(quad_ratio / texture_ratio, -1.);
+                let offset = (1.0 - scale) * 0.5;
+                *uv_offset_scale = Vec4::new(1.0 - scale.x, offset.y, scale.x, scale.y);
+            }
+        }
+        ScalingMode::FitCenter => {
+            let scale = if texture_ratio > quad_ratio {
+                // Scale based on width
+                Vec2::new(1.0, quad_ratio / texture_ratio)
+            } else {
+                // Scale based on height
+                Vec2::new(texture_ratio / quad_ratio, 1.0)
+            };
+
+            *quad_size *= scale;
+        }
+        ScalingMode::FitStart => {
+            if texture_ratio > quad_ratio {
+                // The quad is scaled to match the image ratio, and the quad translation is adjusted
+                // to start of the quad within the original quad size.
+                let scale = Vec2::new(1.0, quad_ratio / texture_ratio);
+                let new_quad = *quad_size * scale;
+                let offset = *quad_size - new_quad;
+                *quad_translation = Vec2::new(0.0, -offset.y);
+                *quad_size = new_quad;
+            } else {
+                let scale = Vec2::new(texture_ratio / quad_ratio, 1.0);
+                let new_quad = *quad_size * scale;
+                let offset = *quad_size - new_quad;
+                *quad_translation = Vec2::new(offset.x, 0.0);
+                *quad_size = new_quad;
+            }
+        }
+        ScalingMode::FitEnd => {
+            if texture_ratio > quad_ratio {
+                let scale = Vec2::new(1.0, quad_ratio / texture_ratio);
+                let new_quad = *quad_size * scale;
+                let offset = *quad_size - new_quad;
+                *quad_translation = Vec2::new(0.0, offset.y);
+                *quad_size = new_quad;
+            } else {
+                let scale = Vec2::new(texture_ratio / quad_ratio, 1.0);
+                let new_quad = *quad_size * scale;
+                let offset = *quad_size - new_quad;
+                *quad_translation = Vec2::new(-offset.x, 0.0);
+                *quad_size = new_quad;
+            }
+        }
```

Comment:
These are identical except for 1 minus sign, and there are similar duplication in the other Start/End cases. On one hand it would be nice if the repetition could be factored out so e.g. FitEnd is just FitStart and set the minus. On the other hand this code is pretty simple and easy to read.

It might also become simpler if you match on `(scaling_mode, quad_ratio > texture_ratio)` since every case has that if inside.

Reply:
Maybe this API would be better with two enums `Fill/Fit` and `Start/Center/End`.

Reply:
I understand your concern about code duplication, and I agree that maintaining clarity is important. However, considering the minimal amount of code duplication we currently have, I believe it's reasonable to keep things as they are. The existing structure seems to be functioning well without introducing unnecessary complexity.